protocol/payment.go

```diff
@@ -0,0 +1,525 @@
+
+	Body PaymentMessageBody `json:"body,omitempty"`
+
+	From string `json:"from,omitempty"`
+	To   string `json:"to,omitempty"`
+}
+
+// PaymentMessageBody represents the body of the PaymentMessage.
+type PaymentMessageBody struct {
+	Payments []Payment `json:"payments"`
+}
+
+// Payment is a union type for field Payments in PaymentMessageBody.
+// Only one of the fields can be set at a time.
+type Payment struct {
+	dataType string
+	crypto   *Iden3PaymentCryptoV1
+	rails    *Iden3PaymentRailsV1
+	railsERC *Iden3PaymentRailsERC20V1
+}
+
+// NewPaymentCrypto creates a new Payment with Iden3PaymentCryptoV1 data.
+func NewPaymentCrypto(data Iden3PaymentCryptoV1) Payment {
+	return Payment{
+		dataType: Iden3PaymentCryptoV1Type,
+		crypto:   &data,
+	}
+}
+
+// NewPaymentRails creates a new Payment with Iden3PaymentRailsV1 data.
+func NewPaymentRails(data Iden3PaymentRailsV1) Payment {
+	return Payment{
+		dataType: Iden3PaymentRailsV1Type,
+		rails:    &data,
+	}
+}
+
+// NEwPaymentRailsERC20 creates a new Payment with Iden3PaymentRailsERC20V1 data.
+func NEwPaymentRailsERC20(data Iden3PaymentRailsERC20V1) Payment {
+	return Payment{
+		dataType: Iden3PaymentRailsERC20V1Type,
+		railsERC: &data,
+	}
+}
+
+// Type returns the type of the data in the union. You can use Data() to get the data.
+func (p *Payment) Type() string {
+	return p.dataType
+}
+
+// Data returns the data in the union. You can use Type() to determine the type of the data.
+func (p *Payment) Data() interface{} {
+	switch p.dataType {
+	case Iden3PaymentCryptoV1Type:
+		return p.crypto
+	case Iden3PaymentRailsV1Type:
+		return p.rails
+	case Iden3PaymentRailsERC20V1Type:
+		return p.railsERC
+	}
+	return nil
+}
+
+// UnmarshalJSON unmarshal the Payment from JSON.
+func (p *Payment) UnmarshalJSON(bytes []byte) error {
+	var item struct {
+		Type string `json:"type"`
+	}
+	if err := json.Unmarshal(bytes, &item); err != nil {
+		return fmt.Errorf("failed to unmarshal Payment: %w", err)
+	}
+
+	p.dataType = item.Type
+	switch item.Type {
+	case Iden3PaymentCryptoV1Type:
+		return json.Unmarshal(bytes, &p.crypto)
+	case Iden3PaymentRailsV1Type:
+		return json.Unmarshal(bytes, &p.rails)
+	case Iden3PaymentRailsERC20V1Type:
+		return json.Unmarshal(bytes, &p.railsERC)
+	}
+	return errors.Errorf("failed to unmarshal PaymentRequestInfoData: %s", string(bytes))
+}
+
+// MarshalJSON marshals the Payment into JSON.
+func (p Payment) MarshalJSON() ([]byte, error) {
+	switch p.dataType {
+	case Iden3PaymentCryptoV1Type:
+		return json.Marshal(p.crypto)
+	case Iden3PaymentRailsV1Type:
+		return json.Marshal(p.rails)
+	case Iden3PaymentRailsERC20V1Type:
+		return json.Marshal(p.railsERC)
+	}
+	return nil, errors.New("failed to marshal not initialized Payment")
+}
+
+// Iden3PaymentCryptoV1 represents the Iden3PaymentCryptoV1 payment data.
+type Iden3PaymentCryptoV1 struct {
+	ID          string         `json:"id"`
+	Type        string         `json:"type"`
```

Comment:
Could we define separate type - iden3comm.PaymentType or similar ? 
For this file `Type` only one value are possible - `Iden3PaymentRequestCryptoV1`. 
The same for  Iden3PaymentRailsV1 and Iden3PaymentRailsERC20V1

Reply:
Done.